### PR TITLE
Move types from progressreader and broadcastwriter to broadcaster

### PR DESCRIPTION
**Pull request by tiborvass**
progressreader.Broadcaster becomes broadcaster.Buffered and
broadcastwriter.Writer becomes broadcaster.Unbuffered.

The package broadcastwriter is thus renamed to broadcaster.

Signed-off-by: Tibor Vass tibor@docker.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
unit test failures

**Comment by icecrime** [comment]
No strong opinion here on my end.

Ping @aaronlehmann @LK4D4 @unclejack: LGTY?

File: pkg/broadcaster/unbuffered.go [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
+package broadcaster
 
 import (
 	"io"
 	"sync"
 )
 
-// BroadcastWriter accumulate multiple io.WriteCloser by stream.
-type BroadcastWriter struct {
+// Unbuffered accumulate multiple io.WriteCloser by stream.
```
Comment (aaronlehmann) [inline pkg/broadcaster/unbuffered.go]: Not related to this change, but should s/accumulate/accumulates/.

File: pkg/broadcaster/unbuffered.go [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
+type Unbuffered struct {
 	mu      sync.Mutex
 	writers []io.WriteCloser
 }
 
 // AddWriter adds new io.WriteCloser.
-func (w *BroadcastWriter) AddWriter(writer io.WriteCloser) {
+func (w *Unbuffered) AddWriter(writer io.WriteCloser) {
```
Comment (aaronlehmann) [inline pkg/broadcaster/unbuffered.go]: I suggest renaming this to `Add` for consistency with the buffered version.

This version takes a `io.WriteCloser`, but the buffered version takes an `io.Writer`. I don't see any obvious way to harmonize this, though.

**Comment by aaronlehmann** [comment]
LGTM. Noted some very minor nits.

File: pkg/broadcaster/unbuffered.go (hunk 2 above)
Comment (tiborvass) [inline pkg/broadcaster/unbuffered.go]: @aaronlehmann we could make it an `io.Writer` and try casting it to `io.Closer` wherever we need to call `Close()` how does that sound?

Comment (aaronlehmann) [inline pkg/broadcaster/unbuffered.go]: I think I'd rather not. If the writers we're passing in have always have a `Close` method and we call that `Close` method, I think it's better to type them as `io.WriteClosers`. The difference between the buffered and unbuffered versions isn't ideal, but I'd rather have that different plainly exposed than disguise it.

Comment (tiborvass) [inline pkg/broadcaster/unbuffered.go]: ok

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**